<a href="https://colab.research.google.com/github/keishauliiaa-prog/PraktikDasarStatistikPython/blob/main/Latihan_Exploratory_Data_Analysis_(EDA)_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🧭 Exploratory Data Analysis (EDA)

### 🚀 Tujuan
*   Membuat pertanyaan analisis atau bisnis yang ingin dicari jawabannya

*   Melakukan eksplorasi terhadap setiap data untuk mencari insight menarik guna menjawab pertanyaan bisnis tersebut.

### 🔄 Alur Latihan
1. Tahap penentuan pertanyaan bisnis untuk explorasi data.
2. Tahap persiapan data.
3. Tahap eksplorasi data.

### 📊 Data Sheet : [Kaggle - Shopping Cart](https://www.kaggle.com/datasets/ruchi798/shopping-cart-database)

# 1️⃣ Tahap penentuan pertanyaan bisnis untuk explorasi data

Sebagai perusahaan yang bergerak di bidang online fashion, kita ingin memahami pelanggan, penjualan, dan performa operasional untuk meningkatkan strategi penjualan dan efisiensi bisnis.
Dataset yang digunakan meliputi:

* `customers.csv` → Data pelanggan

* `orders.csv` → Data pesanan

* `products.csv` → Data produk fashion

* `sales.csv` → Data transaksi per produk dalam pesanan


|  No | Fokus Analisis                                        | Pertanyaan Utama                                                                                       | Tujuan Bisnis                                                                        |
| :-: | :---------------------------------------------------- | :----------------------------------------------------------------------------------------------------- | :----------------------------------------------------------------------------------- |
| 1️⃣ | 🧍‍♀️ **Demografi Pelanggan (Customer Demographics)** | Siapa segmen pelanggan utama kita berdasarkan usia, jenis kelamin, dan lokasi?                         | Menentukan target audiens paling potensial untuk kampanye iklan yang lebih efisien.  |
| 2️⃣ | 🛍️ **Aktivitas Order Pelanggan**                     | Berapa banyak pelanggan aktif yang rutin melakukan pembelian, dan seberapa sering mereka bertransaksi? | Mengukur loyalitas pelanggan serta menentukan strategi retensi (customer retention). |
| 3️⃣ | 🌏 **Demografi Persebaran Data Order**                | Wilayah geografis mana yang memiliki frekuensi pembelian dan nilai transaksi tertinggi?                | Menentukan area prioritas untuk ekspansi pemasaran atau distribusi.                  |
| 4️⃣ | 👕 **Demografi Persebaran Data Produk**               | Produk jenis apa (tipe, warna, ukuran, atau harga) yang paling laku di pasaran?                        | Mengidentifikasi produk unggulan dan menentukan strategi stok serta pricing.         |
| 5️⃣ | 📈 **Tren Penjualan Setiap Bulan**                    | Bagaimana pola penjualan setiap bulan, dan kapan periode puncak penjualan terjadi?                     | Membantu dalam perencanaan stok dan strategi promosi musiman agar lebih efektif.     |


# 2️⃣ Tahap persiapan data

In [ ]:
# Import library
import numpy as np                  # Untuk perhitungan numerik
import pandas as pd                 # Untuk membuat dan mengelola DataFrame
import matplotlib.pyplot as plt     # Untuk visualisasi dasar
import seaborn as sns               # Untuk visualisasi lanjutan

# Koneksi ke Google Drive
from google.colab import drive
drive.mount('/content/drive')       # Memasukkan Google Drive agar file dapat diakses

Mounted at /content/drive


🔹 Keterangan:

* `drive.mount('/content/drive')` → menghubungkan Colab dengan Google Drive agar file CSV bisa diakses.

* Perlu dilakukan sekali per sesi jika ingin membaca file dari Drive.

In [ ]:
# Membuat variabel file path
file_customers = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 5 - Exploratory Data Analysis (EDA)/Modul Praktik/customers_cleaned.csv'  # File customers
file_orders = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 5 - Exploratory Data Analysis (EDA)/Modul Praktik/orders_cleaned.csv'        # File orders
file_products = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 5 - Exploratory Data Analysis (EDA)/Modul Praktik/products_cleaned.csv'    # File products
file_sales = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 5 - Exploratory Data Analysis (EDA)/Modul Praktik/sales_cleaned.csv'          # File sales


# Membaca data dari file CSV ke dalam DataFrame
customers_df = pd.read_csv(file_customers)   # Membaca data pelanggan
orders_df = pd.read_csv(file_orders)         # Membaca data order
products_df = pd.read_csv(file_products)     # Membaca data produk
sales_df = pd.read_csv(file_sales)           # Membaca data penjualan

# Menampilkan 5 baris pertama setiap DataFrame untuk melihat isi data
print(customers_df.head())    # 5 data pertama pelanggan
print(orders_df.head())       # 5 data pertama order
print(products_df.head())     # 5 data pertama produk
print(sales_df.head())        # 5 data pertama penjualan


   customer_id customer_name  gender  age                    home_address  \
0            1        Nama 1  Female   30  8606 Victoria TerraceSuite 560   
1            2        Nama 2  Female   69      8327 Kirlin SummitApt. 461   
2            3        Nama 3  Female   59       269 Gemma SummitSuite 109   
3            4        Nama 4  Female   67       743 Bailey GroveSuite 141   
4            5        Nama 5  Female   30         48 Hyatt ManorSuite 375   

   zip_code               city                         state    country  
0      5464      Johnstonhaven            Northern Territory  Australia  
1      8223    New Zacharyfort               South Australia  Australia  
2      5661           Aliburgh  Australian Capital Territory  Australia  
3      1729  South Justinhaven                    Queensland  Australia  
4      4032     Griffithsshire                    Queensland  Australia  
   order_id  customer_id  payment  order_date delivery_date
0         1           64    30811

🔹 Keterangan:

* `pd.read_csv(file_path)` → membaca file CSV menjadi DataFrame pandas.

* `DataFrame` memudahkan manipulasi data seperti filtering, agregasi, dan analisis statistik.

* `df.head()` → menampilkan beberapa baris pertama, default 5 baris.

# 3️⃣ Tahap eksplorasi data

### 🧍‍♀️ Demografi Pelanggan (Customer Demographics)

Tujuan: Mengetahui siapa pelanggan utama kita dan di mana mereka berada untuk mendukung targeted marketing atau ads campaign.

In [ ]:
# Distribusi Jenis Kelamin

# Menghitung jumlah masing-masing kategori gender (misal: Male/Female)
gender_summary = customers_df['gender'].value_counts().reset_index()   # Mengembalikan jumlah kemunculan tiap nilai unik dalam kolom 'gender', lalu ubah jadi DataFrame

gender_summary.columns = ['Gender', 'Jumlah']   # Mengganti nama kolom agar lebih mudah dibaca: 'index' → 'Gender', 'gender' → 'Jumlah'

# Menambahkan kolom persentase dari total keseluruhan
gender_summary['Persentase (%)'] = round(      # Membulatkan hasil perhitungan ke dua angka di belakang koma
    (gender_summary['Jumlah'] / gender_summary['Jumlah'].sum()) * 100,  # Rumus: (jumlah per gender / total semua gender) * 100
    2
)

print("Distribusi Jenis Kelamin")  # Menampilkan judul output
print(gender_summary)              # Menampilkan tabel hasil analisis distribusi gender

Distribusi Jenis Kelamin
              Gender  Jumlah  Persentase (%)
0             Female     514            51.4
1               Male     437            43.7
2  Prefer not to say      49             4.9


In [ ]:
# Statistik Usia

print("Statistik Usia Pelanggan")   # Menampilkan judul bagian analisis

# Menampilkan statistik deskriptif dari kolom 'age'
print(customers_df['age'].describe())  # Fungsi describe() memberikan ringkasan statistik:
                                       # - count   : jumlah data usia yang valid (tidak null)
                                       # - mean    : rata-rata usia
                                       # - std     : standar deviasi (tingkat penyebaran usia)
                                       # - min     : usia minimum
                                       # - 25%     : kuartil pertama (batas bawah 25% data)
                                       # - 50%     : median (titik tengah data)
                                       # - 75%     : kuartil ketiga (batas atas 75% data)
                                       # - max     : usia maksimum

Statistik Usia Pelanggan
count    1000.000000
mean       49.860000
std        17.647828
min        20.000000
25%        34.000000
50%        50.000000
75%        65.000000
max        80.000000
Name: age, dtype: float64


In [ ]:
# Distribusi Kelompok Usia

# Menentukan batas (range) usia untuk pengelompokan
bins = [0, 20, 30, 40, 50, 60, 70, 100]      # Rentang interval usia (batas bawah dan atas setiap kelompok)
labels = ['<20', '21-30', '31-40', '41-50',  # Label yang merepresentasikan setiap rentang usia
          '51-60', '61-70', '70+']

# Membuat kolom baru 'age_group' berdasarkan pembagian usia menggunakan pd.cut()
customers_df['age_group'] = pd.cut(
    customers_df['age'],                     # Kolom usia yang akan dikelompokkan
    bins=bins,                               # Daftar rentang usia
    labels=labels,                           # Label untuk setiap kelompok
    right=False                              # Interval kiri-tertutup, kanan-terbuka (misal: 20 masuk ke 20–30, bukan <20)
)

# Menghitung jumlah pelanggan di setiap kelompok usia
age_group_summary = customers_df['age_group'].value_counts()  # Hitung jumlah tiap kategori 'age_group'
age_group_summary = age_group_summary.sort_index()            # Urutkan sesuai urutan label (bukan abjad)
age_group_summary = age_group_summary.reset_index()           # Ubah jadi DataFrame agar bisa diolah lebih lanjut

age_group_summary.columns = ['Kelompok Usia', 'Jumlah']       # Ganti nama kolom agar lebih deskriptif

# Hitung persentase setiap kelompok terhadap total
age_group_summary['Persentase (%)'] = round(
    (age_group_summary['Jumlah'] / age_group_summary['Jumlah'].sum()) * 100, 2
)

print("Distribusi Berdasarkan Kelompok Usia")  # Tampilkan judul
print(age_group_summary)                       # Tampilkan tabel hasil distribusi usia

Distribusi Berdasarkan Kelompok Usia
  Kelompok Usia  Jumlah  Persentase (%)
0           <20       0             0.0
1         21-30     164            16.4
2         31-40     167            16.7
3         41-50     160            16.0
4         51-60     166            16.6
5         61-70     169            16.9
6           70+     174            17.4


In [ ]:
# Distribusi Negara

# Menghitung jumlah pelanggan berdasarkan negara
country_summary = customers_df['country'].value_counts().reset_index()   # Hitung jumlah data untuk tiap negara, lalu ubah hasil Series jadi DataFrame

country_summary.columns = ['Negara', 'Jumlah']   # Ganti nama kolom agar lebih informatif

# Hitung persentase tiap negara terhadap total pelanggan
country_summary['Persentase (%)'] = round(
    (country_summary['Jumlah'] / country_summary['Jumlah'].sum()) * 100,  # Rumus: (jumlah per negara / total pelanggan) * 100
    2                                                                     # Hasil dibulatkan ke 2 angka desimal
)

print("Distribusi Berdasarkan Negara")   # Menampilkan judul output
print(country_summary)                   # Menampilkan tabel hasil distribusi pelanggan berdasarkan negara

Distribusi Berdasarkan Negara
      Negara  Jumlah  Persentase (%)
0  Australia    1000           100.0


In [ ]:
# State Teratas

# Menghitung jumlah pelanggan berdasarkan 'state' dan menampilkan hanya 10 teratas
state_summary = customers_df['state'].value_counts()   # Hitung jumlah data untuk tiap state (provinsi/wilayah)
state_summary = state_summary.head(10)                 # Ambil hanya 10 state dengan jumlah pelanggan terbanyak
state_summary = state_summary.reset_index()            # Ubah hasil Series menjadi DataFrame agar mudah diolah

state_summary.columns = ['State', 'Jumlah']            # Ganti nama kolom agar lebih mudah dibaca

print("10 State dengan Jumlah Pelanggan Terbanyak")    # Tampilkan judul analisis
print(state_summary)                                   # Tampilkan tabel hasil

10 State dengan Jumlah Pelanggan Terbanyak
                          State  Jumlah
0               South Australia     139
1                    Queensland     134
2               New South Wales     132
3            Northern Territory     125
4             Western Australia     124
5  Australian Capital Territory     121
6                      Victoria     121
7                      Tasmania     104


In [ ]:
# Kota Teratas

# Menghitung jumlah pelanggan berdasarkan kota dan mengambil 10 teratas
city_summary = customers_df['city'].value_counts()     # Hitung jumlah kemunculan tiap kota (berapa banyak pelanggan di tiap kota)
city_summary = city_summary.head(10)                   # Ambil hanya 10 kota dengan jumlah pelanggan terbanyak
city_summary = city_summary.reset_index()              # Ubah hasil Series jadi DataFrame agar mudah diolah dan diberi nama kolom

city_summary.columns = ['Kota', 'Jumlah']              # Ganti nama kolom agar lebih mudah dibaca

print("10 Kota dengan Jumlah Pelanggan Terbanyak")     # Tampilkan judul hasil analisis
print(city_summary)                                    # Tampilkan tabel hasil top 10 kota

10 Kota dengan Jumlah Pelanggan Terbanyak
                 Kota  Jumlah
0         East Sophia       3
1             New Ava       3
2          East Aidan       3
3           Lake Rose       2
4            East Max       2
5  West Sebastianfort       2
6         Port Hunter       2
7        Williammouth       2
8            Ryanberg       2
9      Schroedermouth       2


🔹 Keterangan:

* `customers_df['gender'].value_counts()` → digunakan untuk menghitung distribusi jumlah pelanggan berdasarkan jenis kelamin.

* `customers_df['age'].describe()` → memberikan ringkasan statistik deskriptif usia pelanggan (mean, min, max, dan quartile).

* `pd.cut()` → digunakan untuk mengelompokkan usia ke dalam kategori umur tertentu.

* `value_counts()` pada kolom state dan city → menunjukkan daerah dengan jumlah pelanggan terbanyak.

* Semua pelanggan berasal dari satu negara: `Australia`, sehingga analisis wilayah difokuskan ke level state dan city.

🔹 Insight Penting:

* Dominasi gender:
Pelanggan perempuan mendominasi dengan `51,4%`, diikuti laki-laki `43,7%`, dan `4,9%` memilih tidak menyebutkan gender.
> Artinya strategi pemasaran bisa difokuskan pada preferensi fashion wanita tanpa mengabaikan pasar pria.

* Distribusi usia merata:
Tidak ada kelompok usia yang menonjol secara signifikan — tiap kelompok (`21–70 tahun`) memiliki proporsi sekitar `16–17%`.
> Produk dan kampanye iklan dapat dibuat lebih universal, mencakup semua kelompok usia dewasa.

* Wilayah utama pelanggan:
Seluruh pelanggan berasal dari Australia, dengan top 3 state adalah: South Australia (`13.9%`), Queensland (`13.4%`), dan New South Wales (`13.2%`)
>  Daerah-daerah ini bisa menjadi target utama untuk distribusi produk atau promosi lokal.

* Kota dengan pelanggan terbanyak:
Kota seperti East Sophia, New Ava, dan East Aidan memiliki pelanggan paling banyak (`masing-masing 3 orang`).
> Walau skalanya kecil, ini bisa jadi indikasi area potensial untuk memperluas jangkauan pemasaran.

### 🛍️ Aktivitas Order Pelanggan
Tujuan: Mengukur seberapa banyak pelanggan yang benar-benar aktif melakukan pembelian dan seberapa sering mereka bertransaksi.

In [ ]:
import pandas as pd   # Mengimpor library pandas untuk manipulasi data

# Menggabungkan Data Order dengan Data Customer
orders_geo = pd.merge(                # Fungsi merge() digunakan untuk menggabungkan dua DataFrame
    orders_df,                        # DataFrame pertama → berisi data pesanan (orders)
    customers_df,                     # DataFrame kedua → berisi data pelanggan (customers)
    on='customer_id',                 # Kolom kunci penghubung di kedua tabel (foreign key)
    how='left'                        # Jenis join: LEFT JOIN → semua data dari orders_df dipertahankan,
                                      # dan data customers_df ditambahkan jika ada customer_id yang cocok
)

In [ ]:
# Hitung jumlah order per pelanggan
customer_activity = orders_df.groupby('customer_id').agg(   # Mengelompokkan data order berdasarkan customer_id
    jumlah_order=('order_id', 'count'),                     # Hitung banyaknya order yang dilakukan setiap pelanggan
    total_pembayaran=('payment', 'sum')                     # Jumlahkan total pembayaran dari setiap pelanggan
).reset_index()                                             # Kembalikan hasil ke bentuk DataFrame biasa

# Gabungkan hasil agregasi dengan data pelanggan
customer_activity = pd.merge(                               # Menggabungkan data aktivitas dengan informasi pelanggan
    customer_activity,                                       # Data hasil agregasi (jumlah order dan total pembayaran)
    customers_df[['customer_id', 'gender', 'age', 'country']], # Kolom penting dari data pelanggan
    on='customer_id',                                       # Kolom penghubung antara kedua DataFrame
    how='left'                                              # LEFT JOIN → semua pelanggan yang pernah order tetap ada
)

# Statistik umum aktivitas pelanggan
total_customer = customers_df['customer_id'].nunique()      # Hitung total pelanggan unik yang terdaftar
customer_order = customer_activity['customer_id'].nunique() # Hitung jumlah pelanggan yang pernah melakukan order
inactive_customer = total_customer - customer_order         # Hitung pelanggan yang belum pernah melakukan order

print("Statistik Aktivitas Pelanggan")                      # Tampilkan judul analisis
print(f"Total Pelanggan Terdaftar       : {total_customer}")  # Jumlah pelanggan keseluruhan
print(f"Pelanggan yang Pernah Order     : {customer_order}")  # Pelanggan aktif (pernah order)
print(f"Pelanggan Tidak Pernah Order    : {inactive_customer}") # Pelanggan pasif (belum pernah order)
print(f"Persentase Pelanggan Aktif (%)  : {round((customer_order / total_customer) * 100, 2)}%") # Rasio pelanggan aktif

Statistik Aktivitas Pelanggan
Total Pelanggan Terdaftar       : 1000
Pelanggan yang Pernah Order     : 617
Pelanggan Tidak Pernah Order    : 383
Persentase Pelanggan Aktif (%)  : 61.7%


In [ ]:
# Kelompokkan pelanggan berdasarkan aktivitasnya

# Fungsi untuk mengelompokkan pelanggan berdasarkan jumlah transaksi
def kategori_order(jumlah):                    # Parameter 'jumlah' = jumlah_order tiap pelanggan
    if jumlah == 1:                            # Jika hanya 1 kali transaksi
        return '1x Transaksi'
    elif 2 <= jumlah <= 3:                     # Jika 2–3 kali transaksi
        return '2-3x Transaksi'
    elif 4 <= jumlah <= 5:                     # Jika 4–5 kali transaksi
        return '4-5x Transaksi'
    else:                                      # Jika lebih dari 5 kali transaksi
        return '>5x Transaksi'

# Terapkan fungsi kategori_order pada kolom 'jumlah_order'
customer_activity['kategori_aktivitas'] = customer_activity['jumlah_order'].apply(kategori_order)
# → apply() menerapkan fungsi custom ke setiap baris kolom tertentu

# Hitung jumlah pelanggan per kategori aktivitas
aktivitas_summary = customer_activity['kategori_aktivitas'].value_counts().reset_index()
# value_counts() menghitung jumlah pelanggan di tiap kategori
# reset_index() mengubah hasilnya menjadi DataFrame agar mudah diolah

aktivitas_summary.columns = ['Kategori Aktivitas', 'Jumlah Pelanggan']  # Ganti nama kolom agar lebih jelas

# Hitung persentase setiap kategori terhadap total pelanggan
aktivitas_summary['Persentase (%)'] = round(
    (aktivitas_summary['Jumlah Pelanggan'] / aktivitas_summary['Jumlah Pelanggan'].sum()) * 100, 2
)

print("Distribusi Pelanggan Berdasarkan Frekuensi Transaksi")  # Judul hasil analisis
print(aktivitas_summary)                                       # Tampilkan tabel hasil

Distribusi Pelanggan Berdasarkan Frekuensi Transaksi
  Kategori Aktivitas  Jumlah Pelanggan  Persentase (%)
0       1x Transaksi               354           57.37
1     2-3x Transaksi               243           39.38
2     4-5x Transaksi                19            3.08
3      >5x Transaksi                 1            0.16


In [ ]:
# Top 10 Pelanggan dengan Total Pembayaran Tertinggi

# Urutkan pelanggan berdasarkan total pembayaran dari terbesar ke terkecil
top_customers = customer_activity.sort_values(
    by='total_pembayaran',       # Kolom yang dijadikan patokan urutan
    ascending=False              # Descending → pembayaran terbesar di atas
).head(10)                       # Ambil 10 pelanggan teratas

print("10 Pelanggan dengan Total Pembayaran Tertinggi")  # Judul output

# Tampilkan kolom penting saja agar lebih ringkas
print(top_customers[['customer_id', 'jumlah_order', 'total_pembayaran', 'country']])

10 Pelanggan dengan Total Pembayaran Tertinggi
     customer_id  jumlah_order  total_pembayaran    country
410          664             5            224843  Australia
347          566             5            212616  Australia
166          282             5            205614  Australia
351          571             6            194472  Australia
288          465             4            163068  Australia
120          206             4            161126  Australia
320          515             4            154845  Australia
63           100             3            152870  Australia
546          882             4            148594  Australia
196          328             3            148339  Australia


🔹 Keterangan:

* `pd.merge()` → digunakan untuk menggabungkan data pelanggan (`customers_df`) dengan data order (`orders_df`) berdasarkan `customer_id`.

* `groupby('customer_id').agg()` → menghitung jumlah order dan total pembayaran per pelanggan.

* Fungsi `kategori_order()` → mengelompokkan pelanggan berdasarkan frekuensi transaksi (`1x`, `2–3x`, `4–5x`, `>5x`).

* `value_counts()` → menghitung distribusi jumlah pelanggan di tiap kategori aktivitas transaksi.

🔹 Insight Penting:

* Tingkat Aktivitas Pelanggan:
Dari total 1000 pelanggan, sebanyak 617 pelanggan (61,7%) pernah melakukan transaksi, sementara 383 pelanggan (38,3%) belum pernah berbelanja.
> Lebih dari separuh pelanggan aktif, namun masih ada potensi besar untuk mengaktivasi pelanggan pasif.

* Frekuensi Transaksi:
Sebagian besar pelanggan hanya bertransaksi 1 kali (`57,37%`), dan `39,38%` melakukan pembelian 2–3 kali.
>  Menunjukkan loyalitas pelanggan masih rendah, perlu strategi retensi seperti voucher, loyalty point, atau personalisasi promo.

* Pelanggan Premium (High Spender):
Pelanggan dengan ID 664, 566, dan 282 mencatat total pembayaran tertinggi, masing-masing lebih dari `200 ribu AUD` dengan frekuensi `4–6 kali` transaksi.
> Segmen pelanggan ini bisa dijadikan target program loyalitas atau VIP karena kontribusinya besar terhadap total revenue.

* Distribusi Negara:
Seluruh pelanggan berasal dari `Australia`, sehingga pengelompokan aktivitas dapat difokuskan pada level demografi seperti usia dan wilayah bagian (state).

### 🌏 Demografi Persebaran Data Order

Tujuan: Mengetahui lokasi geografis yang memiliki aktivitas pembelian paling tinggi.

In [ ]:
# Jumlah Order per Negara

# Mengelompokkan data order berdasarkan negara dan menghitung jumlah order & total pembayaran
country_orders = (
    orders_geo.groupby('country').agg(
        jumlah_order=('order_id', 'count'),    # Hitung jumlah order untuk tiap negara
        total_pembayaran=('payment', 'sum')   # Jumlahkan total pembayaran untuk tiap negara
    )
    .reset_index()                             # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='jumlah_order', ascending=False)  # Urutkan berdasarkan jumlah order terbesar
)

print("Persebaran Order Berdasarkan Negara")   # Judul output
print(country_orders)                          # Tampilkan tabel hasil

Persebaran Order Berdasarkan Negara
     country  jumlah_order  total_pembayaran
0  Australia          1000          33972936


In [ ]:
# 10 State dengan Jumlah Order Tertinggi

# Mengelompokkan data order berdasarkan state dan menghitung jumlah order & total pembayaran
state_orders = (
    orders_geo.groupby('state').agg(
        jumlah_order=('order_id', 'count'),      # Hitung jumlah order untuk tiap state
        total_pembayaran=('payment', 'sum')     # Jumlahkan total pembayaran untuk tiap state
    )
    .reset_index()                               # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='jumlah_order', ascending=False)  # Urutkan berdasarkan jumlah order terbesar
    .head(10)                                    # Ambil 10 state teratas
)

print("10 State dengan Jumlah Order Tertinggi")  # Judul output
print(state_orders)                               # Tampilkan tabel hasil

10 State dengan Jumlah Order Tertinggi
                          state  jumlah_order  total_pembayaran
4               South Australia           148           4873510
3                    Queensland           139           4903649
7             Western Australia           130           4257966
1               New South Wales           129           4272862
6                      Victoria           118           4023684
0  Australian Capital Territory           118           4013131
5                      Tasmania           112           3871117
2            Northern Territory           106           3757017


In [ ]:
# 10 Kota dengan Jumlah Order Tertinggi

# Mengelompokkan data order berdasarkan kota dan menghitung jumlah order & total pembayaran
city_orders = (
    orders_geo.groupby('city').agg(
        jumlah_order=('order_id', 'count'),      # Hitung jumlah order untuk tiap kota
        total_pembayaran=('payment', 'sum')     # Jumlahkan total pembayaran untuk tiap kota
    )
    .reset_index()                               # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='jumlah_order', ascending=False)  # Urutkan berdasarkan jumlah order terbesar
    .head(10)                                    # Ambil 10 kota teratas
)

print("10 Kota dengan Jumlah Order Tertinggi")   # Judul output
print(city_orders)                               # Tampilkan tabel hasil

10 Kota dengan Jumlah Order Tertinggi
                 city  jumlah_order  total_pembayaran
198        Jordanside             6            194472
310           New Ava             6            159898
413  Port Hannahburgh             5            205614
266         Lake Rose             5            185328
384        O'keefeton             5            212616
105          East Max             5            224843
450          Rubyfort             5            119268
574          West Kai             5            125889
462         Sarahstad             4            109299
451        Russelstad             4            111451


🔹 Keterangan:

* `pd.merge()` → digunakan untuk menggabungkan data order (`orders_df`) dengan data pelanggan (`customers_df`) agar setiap order memiliki informasi lokasi.

* `groupby() + agg()` → menghitung jumlah order (`jumlah_order`) dan total pembayaran (total_pembayaran) berdasarkan wilayah seperti negara, state, dan kota.

* `sort_values()` → mengurutkan hasil berdasarkan jumlah order tertinggi agar mudah mengidentifikasi wilayah dengan aktivitas tertinggi.

* Karena seluruh pelanggan berasal dari Australia, analisis fokus pada state dan city sebagai level geografis pembeda.

🔹 Insight Penting:

* Pusat Aktivitas Utama:
Seluruh transaksi berasal dari Australia dengan total `1.000` order dan nilai pembayaran mencapai `33,9 juta AUD`.
> Menunjukkan pasar masih terfokus di satu negara, namun potensial untuk ekspansi internasional.

* Top 3 State dengan Aktivitas Order Tertinggi: `South Australia` – 148 order (AUD 4,87 juta), `Queensland` – 139 order (AUD 4,90 juta), `Western Australia` – 130 order (AUD 4,25 juta)
> Ketiga wilayah ini menyumbang volume dan pendapatan terbesar — cocok dijadikan prioritas kampanye regional.

* Performa Kota:
Kota `Jordanside` dan `New Ava` mencatat jumlah order tertinggi (`6 order`), sedangkan East Max mencatat nilai transaksi tertinggi (`224.843 AUD`).
> Kota-kota tersebut dapat diprioritaskan untuk pengiriman cepat, stok produk, atau promosi targeted ads.

* Pola Geografis yang Stabil:
Setiap state besar memiliki kontribusi relatif seimbang, menunjukkan distribusi penjualan yang cukup merata di seluruh Australia.
> Strategi pemasaran nasional masih efektif, namun fokus tambahan di kota top-performer bisa meningkatkan efisiensi kampanye.

### 👕 Demografi Persebaran Data Produk

Tujuan: Menganalisis performa produk berdasarkan tipe, warna, ukuran, dan harga untuk mengidentifikasi produk paling populer.

In [ ]:
import pandas as pd   # Mengimpor library pandas untuk manipulasi data

# Gabungkan data produk dengan data penjualan
product_sales = pd.merge(
    sales_df,                   # DataFrame berisi data penjualan
    products_df,                # DataFrame berisi informasi produk
    on='product_id',            # Kolom kunci penghubung di kedua DataFrame
    how='left'                  # LEFT JOIN → semua data penjualan tetap ada, info produk ditambahkan jika cocok
)

In [ ]:
# Pastikan nama kolom 'quantity' benar
# Kadang merge menyebabkan duplikasi kolom (quantity_x dan quantity_y)
if 'quantity_x' in product_sales.columns:
    product_sales.rename(columns={'quantity_x': 'quantity'}, inplace=True)  # Gunakan kolom dari sales_df
elif 'quantity_y' in product_sales.columns:
    product_sales.rename(columns={'quantity_y': 'quantity'}, inplace=True)  # Gunakan kolom dari products_df jika ada

# Hitung total penjualan per produk
product_summary = product_sales.groupby(
    ['product_id', 'product_type', 'product_name', 'size', 'colour']  # Kelompokkan berdasarkan atribut produk
).agg(
    total_terjual=('quantity', 'sum'),            # Total jumlah unit terjual
    total_pendapatan=('total_price', 'sum'),      # Total pendapatan dari penjualan produk
    harga_rata_rata=('price_per_unit', 'mean')    # Rata-rata harga per unit (bisa untuk analisis margin/harga)
).reset_index()                                   # Kembalikan hasil groupby menjadi DataFrame biasa

In [ ]:
# Analisis berdasarkan tipe produk

# Mengelompokkan produk berdasarkan 'product_type' dan menghitung performa
type_summary = (
    product_summary.groupby('product_type').agg(
        total_item_terjual=('total_terjual', 'sum'),   # Jumlah total unit terjual per tipe produk
        total_pendapatan=('total_pendapatan', 'sum'),  # Total pendapatan dari tiap tipe produk
        harga_rata_rata=('harga_rata_rata', 'mean')    # Rata-rata harga per unit per tipe produk
    )
    .reset_index()                                      # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='total_item_terjual', ascending=False)  # Urutkan dari tipe produk yang terjual terbanyak
)

print("Performa Berdasarkan Tipe Produk")  # Judul output
print(type_summary)                        # Tampilkan tabel hasil analisis

Performa Berdasarkan Tipe Produk
  product_type  total_item_terjual  total_pendapatan  harga_rata_rata
2     Trousers                3360            341174       101.851942
0       Jacket                3343            357026       106.516049
1        Shirt                3259            333600       102.201923


In [ ]:
# Analisis berdasarkan warna produk

# Mengelompokkan produk berdasarkan 'colour' dan menghitung performa
colour_summary = (
    product_summary.groupby('colour').agg(
        total_item_terjual=('total_terjual', 'sum'),   # Jumlah total unit terjual per warna
        total_pendapatan=('total_pendapatan', 'sum')   # Total pendapatan dari tiap warna
    )
    .reset_index()                                      # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='total_item_terjual', ascending=False)  # Urutkan berdasarkan jumlah item terjual terbanyak
)

print("Performa Berdasarkan Warna Produk")  # Judul output
print(colour_summary)                       # Tampilkan tabel hasil analisis

Performa Berdasarkan Warna Produk
   colour  total_item_terjual  total_pendapatan
0    blue                1528            158220
5  violet                1448            149960
2  indigo                1445            149439
4     red                1407            145898
6  yellow                1389            144168
1   green                1386            143358
3  orange                1359            140757


In [ ]:
# Analisis berdasarkan ukuran produk

# Mengelompokkan produk berdasarkan 'size' dan menghitung performa
size_summary = (
    product_summary.groupby('size').agg(
        total_item_terjual=('total_terjual', 'sum'),   # Jumlah total unit terjual per ukuran
        total_pendapatan=('total_pendapatan', 'sum')   # Total pendapatan dari tiap ukuran
    )
    .reset_index()                                      # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='total_item_terjual', ascending=False)  # Urutkan berdasarkan jumlah item terjual terbanyak
)

print("Performa Berdasarkan Ukuran Produk")  # Judul output
print(size_summary)                          # Tampilkan tabel hasil analisis

Performa Berdasarkan Ukuran Produk
  size  total_item_terjual  total_pendapatan
1    M                2011            208693
4   XS                1990            207093
2    S                1989            206201
0    L                1972            204049
3   XL                1936            199457
5    l                  29              2792
6    m                  25              2325
7   xl                  10              1190


In [ ]:
# 10 Produk Terpopuler

# Urutkan produk berdasarkan total unit terjual dari terbesar ke terkecil
top_products = product_summary.sort_values(
    by='total_terjual',    # Kolom patokan urutan → jumlah unit terjual
    ascending=False         # Descending → produk terjual terbanyak di atas
).head(10)                  # Ambil 10 produk teratas

print("10 Produk Paling Populer (Berdasarkan Jumlah Penjualan)")  # Judul output

# Tampilkan kolom penting saja agar lebih ringkas
print(top_products[['product_name', 'product_type', 'colour', 'size', 'total_terjual', 'total_pendapatan']])

10 Produk Paling Populer (Berdasarkan Jumlah Penjualan)
       product_name product_type  colour size  total_terjual  total_pendapatan
570          Bomber       Jacket   green   XL             24              2400
76          Flannel        Shirt  orange    L             24              2832
180   Camp Collared        Shirt  orange    M             23              2254
465          Puffer       Jacket   green    M             23              2714
824          Chinos     Trousers     red    L             21              2373
725         Peacoat       Jacket  orange   XS             21              2289
695     Trench Coat       Jacket  orange    M             21              2499
1014    Relaxed Leg     Trousers    blue    L             21              1890
478          Puffer       Jacket  violet    S             20              2360
1162   High-Waisted     Trousers  violet    L             20              1980


🔹 Keterangan:

* `pd.merge(sales_df, products_df, on='product_id')` → digunakan untuk menggabungkan data penjualan dengan informasi produk agar setiap transaksi memiliki detail tipe, warna, ukuran, dan harga.

* `groupby() + agg()` → menghitung metrik utama seperti total item terjual, total pendapatan, dan harga rata-rata per kategori produk.

* Analisis dilakukan pada empat dimensi utama: tipe produk, warna, ukuran, dan nama produk untuk menentukan produk paling populer.

* Terdapat duplikasi kolom quantity pada hasil merge, sehingga dilakukan rename agar analisis tidak error.

* Semua kolom numerik (`quantity`, `total_price`, `price_per_unit`) sudah sesuai tipe data dan siap digunakan untuk perhitungan statistik.

🔹 Insight Penting:

* Performa Tipe Produk:
Trousers menjadi produk paling laku dengan 3.360 item terjual, disusul Jacket (3.343 item) dan Shirt (3.259 item).
> Ketiga kategori ini memiliki performa penjualan yang relatif seimbang — strategi stok dan promosi bisa difokuskan ke ketiganya.

* Performa Warna Produk: Warna blue, violet, dan indigo menempati posisi teratas dalam penjualan, masing-masing dengan lebih dari 1.400 item terjual.
> Konsumen tampak lebih menyukai warna bernuansa dingin, bisa jadi karena kesan netral dan mudah dipadupadankan.

* Performa Ukuran Produk:
Ukuran M, XS, dan S menjadi ukuran dengan penjualan tertinggi — masing-masing terjual sekitar 1.900–2.000 item, disusul oleh ukuran L (1.972 item) dan XL (1.936 item).
> Hal ini menunjukkan bahwa produk dengan ukuran sedang hingga besar (S–XL) memiliki permintaan yang seimbang dan dominan di pasar. Strategi stok dan produksi dapat difokuskan pada ukuran ini karena mewakili mayoritas preferensi pelanggan.

* Top 10 Produk Terlaris:
Produk seperti Bomber Jacket (green, XL) dan Flannel Shirt (orange, L) menjadi produk paling diminati, masing-masing terjual 24 unit.
> Model “Jacket” dan “Shirt” dengan warna mencolok seperti hijau dan oranye menarik minat pelanggan, cocok untuk kampanye visual marketing.

* Harga Rata-Rata Stabil:
Rata-rata harga produk berkisar AUD 100–106, menunjukkan konsistensi strategi harga lintas kategori.
> Harga yang stabil menandakan positioning produk di kelas menengah dan bisa dimanfaatkan untuk strategi diskon musiman tanpa menurunkan persepsi nilai.

### 📈 Tren Penjualan Setiap Bulan

Tujuan: Menganalisis pola dan musim penjualan untuk mendukung perencanaan stok dan strategi kampanye promosi.

In [ ]:
import pandas as pd   # Mengimpor library pandas untuk manipulasi data

# Pastikan kolom 'order_date' dalam format datetime agar bisa diolah sebagai tanggal
orders_df['order_date'] = pd.to_datetime(orders_df['order_date'])

# Gabungkan data penjualan dengan data order untuk mendapatkan tanggal tiap transaksi
sales_orders = pd.merge(
    sales_df,                  # DataFrame berisi data penjualan
    orders_df,                 # DataFrame berisi informasi order (termasuk tanggal)
    on='order_id',             # Kolom penghubung antar DataFrame
    how='left'                 # LEFT JOIN → semua data penjualan tetap ada
)

# Tambahkan kolom bulan dan tahun dari tanggal order
sales_orders['bulan'] = sales_orders['order_date'].dt.month        # Ambil nomor bulan (1–12)
sales_orders['tahun'] = sales_orders['order_date'].dt.year         # Ambil tahun dari tanggal
sales_orders['periode'] = sales_orders['order_date'].dt.to_period('M').astype(str)
# 'periode' = format YYYY-MM → mempermudah agregasi bulanan

In [ ]:
# Hitung total pendapatan dan jumlah order per bulan

monthly_trend = (
    sales_orders.groupby('periode').agg(
        total_pendapatan=('total_price', 'sum'),   # Total pendapatan tiap bulan
        total_order=('order_id', 'nunique')        # Jumlah order unik tiap bulan
    )
    .reset_index()                                  # Kembalikan hasil groupby menjadi DataFrame biasa
    .sort_values(by='periode')                      # Urutkan berdasarkan periode (YYYY-MM)
)

print("Tren Penjualan Setiap Bulan")               # Judul output
print(monthly_trend)                               # Tampilkan tabel hasil

Tren Penjualan Setiap Bulan
   periode  total_pendapatan  total_order
0  2021-01            119333          109
1  2021-02             95080           93
2  2021-03            131364          117
3  2021-04             97530           99
4  2021-05             85597           84
5  2021-06            101458          102
6  2021-07            116081          105
7  2021-08            104565          106
8  2021-09             96526           98
9  2021-10             84266           80


In [ ]:
# Identifikasi bulan dengan performa terbaik dan terendah

# Bulan dengan total pendapatan tertinggi
best_month = monthly_trend.loc[monthly_trend['total_pendapatan'].idxmax()]
# idxmax() → mengembalikan index baris dengan nilai maksimum di kolom 'total_pendapatan'

# Bulan dengan total pendapatan terendah
worst_month = monthly_trend.loc[monthly_trend['total_pendapatan'].idxmin()]
# idxmin() → mengembalikan index baris dengan nilai minimum di kolom 'total_pendapatan'

print("Bulan dengan Penjualan Tertinggi")
print(best_month)                           # Tampilkan bulan dengan pendapatan tertinggi
print("\n")

print("Bulan dengan Penjualan Terendah")
print(worst_month)                          # Tampilkan bulan dengan pendapatan terendah

Bulan dengan Penjualan Tertinggi
periode             2021-03
total_pendapatan     131364
total_order             117
Name: 2, dtype: object


Bulan dengan Penjualan Terendah
periode             2021-10
total_pendapatan      84266
total_order              80
Name: 9, dtype: object


🔹 Keterangan:

* `.loc` → indexer yang digunakan untuk mengakses baris dan kolom DataFrame

* Data tren penjualan diperoleh dengan menggabungkan dataset orders_df dan sales_df, lalu dikelompokkan berdasarkan periode bulan (`periode`).

* Kolom `order_date` telah dikonversi ke format datetime untuk memudahkan ekstraksi bulan dan tahun.

* `total_pendapatan` → total nilai penjualan per bulan.

* `total_order` → jumlah pesanan unik per bulan.

🔹 Insight Penting:

* Penjualan tertinggi terjadi pada Maret 2021, dengan total pendapatan 131.364 dan 117 order, menunjukkan adanya momentum positif (kemungkinan kampanye atau tren musiman).

* Penjualan terendah tercatat pada Oktober 2021, hanya 84.266 pendapatan dengan 80 order, menandakan perlunya strategi promosi tambahan di periode akhir tahun.

* Secara umum, tren penjualan berfluktuasi, namun cenderung meningkat di awal tahun (Q1) dan menurun di pertengahan hingga akhir tahun, sehingga stok dan kampanye promosi sebaiknya difokuskan di kuartal pertama tahun.